### ISSUE TITLE
Imputed weights are lost and can't be restored after saving and loading a project

### ISSUE THREAD

**Issue by cschmidl** [GreenleafLab/ArchR#345]
Dear Jeffrey and Ryan,

whenever I save and then load a project (either by loadArchRProject or readRDS - what is your recommendation to save/load project progress?) I lose the imputed weights. When plotting embeddings it returns "No imputeWeights found, returning NULL", also proj@imputeWeights is empty. 

The imputed weights can't be restored by running addImputeWeights although the function runs with no errors. A folder with the imputed weight files is created in the project dir. Despite no error I attached the the logfile from addImputeWeights anyways.

log file
[ArchR-addImputeWeights-6e285b160c46-Date-2020-10-15_Time-14-56-58.log](https://github.com/GreenleafLab/ArchR/files/5385011/ArchR-addImputeWeights-6e285b160c46-Date-2020-10-15_Time-14-56-58.log)

Thanks a lot for your help!
Chris

**Comment by rcorces** [comment]
I can confirm that I get the exact same behavior. Thanks for reporting this - we will figure it out. Sounds like this may be an issue on both the save and load ends @jgranja24 ?

**Comment by jgranja24** [comment]
Hi @cschmidl, 

I actually designed it to be this way because imputeWeights are tricky to keep track of 
and are extremely fast to compute. We also dont use these weights for any analysis besides
visualization. The main challenge is checking that the imputeWeights hdf5 file matches exactly
with the loaded/saved project vs right now there really isnt a need for that type of checking. Ill
think about this issue more, but I dont see it as a large issue? Sorry for any inconveniences!

Best

Jeff

**Comment by rcorces** [comment]
This makes sense to me but the problem that @cschmidl noticed (and I recapitulated) is that after loading an ArchRProject, there seems to be a problem adding back the impute weights. The folder gets created but using `getImputeWeights` doesnt return the weights. This might be some sort of edge case but at least I was able to recapitulate this with the tutorial data.

**Comment by cschmidl** [comment]
Yes I agree that it's no problem to compute the weights again but at the moment as Ryan pointed out you can't add them back to the project. That makes it practically impossible to continue working on a project after saving and loading, so I would be very grateful for a fix!
Cheers,
Christian

**Comment by rcorces** [comment]
At the moment, you can save your entire R workspace and reload that. I'm
surprised this hasn't come up before which makes me wonder if it is a
problem specific to a certain configuration. Let's see if jeff can
recapitulate it on his end.

**Comment by jgranja24** [comment]
Hmmm I actually never had this issue. When you re-run addImputeWeights it should just overwrite the old ones? 

From `addImputeWeights` it removes the files prior to recreation

```
  if(useHdf5){
    dir.create(file.path(getOutputDirectory(ArchRProj), "ImputeWeights"), showWarnings = FALSE)
    if(randomSuffix){
      weightFiles <- .tempfile("Impute-Weights", tmpdir = file.path(gsub(paste0(getwd(),"/"),"",getOutputDirectory(ArchRProj)), "ImputeWeights"))
      weightFiles <- paste0(weightFiles, "-Rep-", seq_len(nRep))
    }else{
      weightFiles <- file.path(getOutputDirectory(ArchRProj), "ImputeWeights", paste0("Impute-Weights-Rep-", seq_len(nRep)))
    }
  }
  o <- suppressWarnings(file.remove(weightFiles))
```

**Comment by cschmidl** [comment]
Yes it does overwrite the files in the folder, but the weights are not added to the project (the proj@imputeWeights
remains a List of length 0). I think the assigning the new weights to the project is the problem.

**Comment by rcorces** [comment]
My apologies. I made a mistake in my code when I quickly recapitulated this problem (I blame confirmation bias!). I ***cannot*** recapitulate this error.

@cschmidl - perhaps you made the same mistake that I made. I forgot to store the output of `addImputeWeights()` into the `ArchRProject`.

```
> projHeme3 <- loadArchRProject("Save-ProjHeme2/")
Successfully loaded ArchRProject!

                                                   / |
                                                 /    \
            .                                  /      |.
            \\\                              /        |.
              \\\                          /           `|.
                \\\                      /              |.
                  \                    /                |\
                  \\#####\           /                  ||
                ==###########>      /                   ||
                 \\##==......\    /                     ||
            ______ =       =|__ /__                     ||      \\\
        ,--' ,----`-,__ ___/'  --,-`-===================##========>
       \               '        ##_______ _____ ,--,__,=##,__   ///
        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/
        -,____,---'       \\####\\________________,--\\_##,/
           ___      .______        ______  __    __  .______      
          /   \     |   _  \      /      ||  |  |  | |   _  \     
         /  ^  \    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    
        /  /_\  \   |      /     |  |     |   __   | |      /     
       /  _____  \  |  |\  \\___ |  `----.|  |  |  | |  |\  \\___.
      /__/     \__\ | _| `._____| \______||__|  |__| | _| `._____|
    
> getImputeWeights(projHeme3)
Getting ImputeWeights
No imputeWeights found, returning NULL
NULL
> projHeme3 <- addImputeWeights(projHeme3)
ArchR logging to : ArchRLogs/ArchR-addImputeWeights-16113a1c0b7fa9-Date-2020-10-16_Time-11-56-49.log
If there is an issue, please report to github with logFile!
2020-10-16 11:56:50 : Computing Impute Weights Using Magic (Cell 2018), 0 mins elapsed.
> getImputeWeights(projHeme3)
Getting ImputeWeights
List of length 2
names(2): Weights Params
> 
```

**Comment by cschmidl** [comment]
I made the same mistake and can add the weights! However I checked my code and it appears that I was using the <- operator only at the beginning of the project. Later I was working with a subset of the data with new clusters and LSI/Harmony, ran the  imputeweights again without the assign operator and the embeddings looked perfectly smooth - I can't explain that...Anyway, it works!
Thanks for the help!
